docs/architecture/adr-058-authz-hooks.md

```diff
@@ -0,0 +1,141 @@
+# ADR 058: Authz hooks
+
+## Changelog
+
+* 2022-08-02: Initial Draft
+
+## Status
+
+DRAFT Not Implemented
+
+## Abstract
+
+In order to extend the use cases covered by the `x/authz` module, we propose to implement a hook system so that external modules can perform stateful checks and/or state changes while handling a `MsgExec`. 
+
+## Context
+
+Currently, when handling `MsgExec` it is not possible to execute stateful checks before accepting an authorization, nor it is possible to perform state changes after it has been accepted. This makes it impossible to correctly implement some use cases like fee-based authorizations (as described inside [\#11583](https://github.com/cosmos/cosmos-sdk/issues/11583)).
+
+## Decision
+
+In order to improve the number of use cases that can be handled though the usage of `x/authz`, we proposed to implement a hook system within the authz keeper, similar to what other modules already use. 
+
+Particularly, we have thought about the following hook structure: 
+
+```go
+type AuthzHooks interface {
+    AfterAuthorizationAccepted(ctx sdk.Context, grantee sdk.AccAddress, authorization Authorization) error
+}
+```
+
+This can then be used as follows within the `x/auth` keeper structure: 
+
+```go
+type Keeper struct {
+    hooks authz.AuthzHooks
+    // ...
+}
+
+// OnAuthorizationAccepted must be called after an authorization has been accepted
+func (k Keeper) OnAuthorizationAccepted(ctx sdk.Context, authorization authz.Authorization) error {
+    if k.hooks != nil {
+        return k.hooks.OnAuthorizationAccepted(ctx, authorization)
+    }
+    return nil
+}
+
+// DispatchActions attempts to execute the provided messages via authorization
+// grants from the message signer to the grantee.
+func (k Keeper) DispatchActions(ctx sdk.Context, grantee sdk.AccAddress, msgs []sdk.Msg) ([][]byte, error) {
+    // [...]
+
+        authorization, err := grant.GetAuthorization()
+        if err != nil {
+            return nil, err
+        }
+        
+        resp, err := authorization.Accept(ctx, msg)
+        if err != nil {
+            return nil, err
+        }
+        
+        if resp.Delete {
+            err = k.DeleteGrant(ctx, grantee, granter, sdk.MsgTypeURL(msg))
+        } else if resp.Updated != nil {
+            err = k.update(ctx, grantee, granter, resp.Updated)
+        }
+        if err != nil {
+            return nil, err
+        }
+        
+        if !resp.Accept {
+            return nil, sdkerrors.ErrUnauthorized
+        }
```

Comment:
nit: no need to copy the entire structure here, we know that logic comes before using `...`


```suggestion
```